## Preprocessing

In [21]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [22]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(["EIN", "NAME"], axis=1)
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [23]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [24]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
value_counts = application_df['APPLICATION_TYPE'].value_counts()

print(value_counts)

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [25]:
# Choose a cutoff value and create a list of application types to be replaced (Cut-off value is 500)
# use the variable name `application_types_to_replace`

application_types_to_replace = value_counts[value_counts < 500].index

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [26]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_value_counts = application_df['CLASSIFICATION'].value_counts()

print(classification_value_counts)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [27]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_value_counts[classification_value_counts>1]

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64

In [28]:
# Choose a cutoff value and create a list of classifications to be replaced (Cut-off value is 100)
# use the variable name `classifications_to_replace`
classifications_to_replace = classification_value_counts[classification_value_counts < 100].index

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
Other      669
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: count, dtype: int64

In [29]:
removed_field = 'INCOME_AMT'

In [30]:
# Dropping Individual Columns to Test

application_df = application_df.drop(removed_field, axis=1)

In [31]:
categorical_columns = ['APPLICATION_TYPE',
                       'AFFILIATION',
                       'CLASSIFICATION',
                       'USE_CASE',
                       'ORGANIZATION',
                       #'INCOME_AMT'
                       'SPECIAL_CONSIDERATIONS']

In [32]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df_dummies = pd.get_dummies(application_df, columns=categorical_columns)

In [33]:
# Split our preprocessed data into our features and target arrays
y = application_df_dummies['IS_SUCCESSFUL']
X = application_df_dummies.drop(columns=['IS_SUCCESSFUL'])

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=70)

In [34]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [35]:
print(f"There are {len(X_train_scaled[0])} features")
input_size = len(X_train_scaled[0])

There are 40 features


## Compile, Train and Evaluate the Model

In [36]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
nn_model = tf.keras.models.Sequential()

# First hidden layer
nn_model.add(tf.keras.layers.Dense(units = 36, activation='relu', input_dim=input_size))

# Second hidden layer
nn_model.add(tf.keras.layers.Dense(units = 16, activation='relu'))

# Output layer
nn_model.add(tf.keras.layers.Dense(units = 1, activation='sigmoid'))

# Check the structure of the model
nn_model.summary()

c:\Users\tokar\anaconda3\envs\dev\lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 36)             │         1,476 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 16)             │           592 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            17 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,085 (8.14 KB)

 Trainable params: 2,085 (8.14 KB)

 Non-trainable params: 0 (0.00 B)

In [37]:
# Compile the model
nn_model.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [38]:
# Train the model
fit_model = nn_model.fit(X_train_scaled, y_train, validation_split=.15, epochs=100)

Epoch 1/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 907us/step - accuracy: 0.6902 - loss: 0.6020 - val_accuracy: 0.7323 - val_loss: 0.5617
Epoch 2/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 716us/step - accuracy: 0.7227 - loss: 0.5687 - val_accuracy: 0.7308 - val_loss: 0.5602
Epoch 3/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 718us/step - accuracy: 0.7193 - loss: 0.5651 - val_accuracy: 0.7328 - val_loss: 0.5546
Epoch 4/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 726us/step - accuracy: 0.7322 - loss: 0.5549 - val_accuracy: 0.7362 - val_loss: 0.5549
Epoch 5/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 744us/step - accuracy: 0.7275 - loss: 0.5580 - val_accuracy: 0.7323 - val_loss: 0.5543
Epoch 6/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 709us/step - accuracy: 0.7279 - loss: 0.5566 - val_accuracy: 0.7289 - val_loss: 0.5567
Epoch 7/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 709us/step - accuracy: 0.7203 - loss: 0.5602 - val_accuracy: 0.7344 - val_loss: 0.5530
Epoch 8/100
684/684 ━━━━━━━━━━━━━━━━━━━━ 1s 720us/step - accuracy: 0.7278 - loss: 0

In [39]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 512us/step - accuracy: 0.7242 - loss: 0.5584
Loss: 0.5583711862564087, Accuracy: 0.7241982221603394


#### Results dropping APPLICATION_TYPE column

Accuracy: 0.7029 - Loss: 0.5817


#### Results dropping AFFILIATION column

Accuracy: 0.6480 - Loss: 0.6123

#### Results dropping CLASSIFICATION column

Accuracy: 0.7252 - Loss: 0.5737

#### Results dropping USE_CASE column

Accuracy: 0.7243 - Loss: 0.5573

#### Results dropping ORGANIZATION column

Accuracy: 0.7249 - Loss: 0.5684

#### Results dropping INCOME_AMT column

Accuracy: 0.7249 - Loss: 0.5586

#### Results dropping SPECIAL_CONSIDERATIONS column

Accuracy: 0.7261 - Loss: 0.5566

In [40]:
# Export our model to HDF5 file
nn_model.save(f"AlphabetSoupCharity_drop_{removed_field}.h5")